plugins/grpc-ecosystem/gateway/v2.15.2/separate_pkg_additional_imports.patch

```diff
@@ -0,0 +1,291 @@
++
++	// separatePackage determines whether to output the generated code into a separate package.
++	separatePackage bool
++
++	// additionalImports is a list of additional imports to be added to the generated code.
++	// N.B. additional imports is not a flag option
++	additionalImports []string
+ }
+
+ type repeatedFieldSeparator struct {
+@@ -236,7 +243,9 @@ func (r *Registry) loadFile(filePath string, file *protogen.File) {
+ 	if r.standalone {
+ 		pkg.Alias = "ext" + cases.Title(language.AmericanEnglish).String(pkg.Name)
+ 	}
+-
++	if r.separatePackage {
++		pkg.Name += "gateway"
++	}
+ 	if err := r.ReserveGoPackageAlias(pkg.Name, pkg.Path); err != nil {
+ 		for i := 0; ; i++ {
+ 			alias := fmt.Sprintf("%s_%d", pkg.Name, i)
+@@ -811,3 +820,23 @@ func (r *Registry) SetAllowPatchFeature(allow bool) {
+ func (r *Registry) GetAllowPatchFeature() bool {
+ 	return r.allowPatchFeature
+ }
++
++// SetSeparatePackage sets separatePackage
++func (r *Registry) SetSeparatePackage(use bool) {
++	r.separatePackage = use
++}
++
++// GetSeparatePackage returns separatePackage
++func (r *Registry) GetSeparatePackage() bool {
++	return r.separatePackage
++}
++
++// SetAdditionalImports sets additionalImports
++func (r *Registry) SetAdditionalImports(imports []string) {
++	r.additionalImports = imports
++}
++
++// GetAdditionalImports returns additionalImports
++func (r *Registry) GetAdditionalImports() []string {
++	return r.additionalImports
++}
+diff --git a/internal/descriptor/services.go b/internal/descriptor/services.go
+index 8936a782..1fd4d8dc 100644
+--- a/internal/descriptor/services.go
++++ b/internal/descriptor/services.go
+@@ -3,6 +3,7 @@ package descriptor
+ import (
+ 	"errors"
+ 	"fmt"
++	"path/filepath"
+ 	"strings"
+
+ 	"github.com/golang/glog"
+@@ -25,6 +26,31 @@ func (r *Registry) loadServices(file *File) error {
+ 			ServiceDescriptorProto: sd,
+ 			ForcePrefixedName:      r.standalone,
+ 		}
++		if r.separatePackage {
++			// when generating a separate package for the gateway, we need to generate an import statement
++			// for the gRPC stubs that are no longer in the same package. This is done by adding the grpc
++			// package to the additionalImports list. In order to prepare a valid import statement, we'll replace
++			// the source package name, something like: ../pet/v1/v1petgateway with ../pet/v1/v1petgrpc
++			const (
++				baseTypePackageName = "protocolbuffers"
++				grpcPackageName     = "grpc"
++			)
++			packageName := strings.ReplaceAll(svc.File.GoPkg.Name, "gateway", "") + grpcPackageName
++			svc.GRPCFile = &File{
++				GoPkg: GoPackage{
++					// additionally, as the `go_package` option is passed through from the generator, and can only be
++					// set the one time, without making major changes, we'll use the package name sent through the
++					// options as a basis, and replace the source package name with the grpc package name.
++					Path: strings.ReplaceAll(
++						filepath.Join(svc.File.GoPkg.Path, packageName),
```

Comment:
Similarly - this shouldn't use ReplaceAll - I think we want to replace `/gen/go/protocolbuffers/go/` with `/gen/go/grpc/go/` here or something more specific.

Reply:
You're right, the only caveat is that the module name is placed between `/gen/go` & `/protocolbuffers/go`.
I'll update this to be more specific